### PR TITLE
Added event quest support, 6 Tri event quests, and 1 custom event quest

### PR DESCRIPTION
**Pull request by kbarkevich**
Updated event quest generator to reflect new information

Added quest location enum and foundation for event quest generator functions

Added monster enum

Added quest rank type enum

added monster quest type maker

Added event quest boilerplate (can't launch)

Functional event quests

Added Sample Event Quests & Added Event Quest Analysis

Added egg transport event quest

Added dummy small monster support

Small monster data can be loaded from plaintext hex files

Added item drop table support

Added debug event quest

Added event quest "Jump Three Jaggi" (Jump Four Jaggi replacement)

Added event quest "Blood Sport"

Updated UNK labels and added "Mercy Mission" event quest

Added accurate World Eater quest

Updated documentation and quest details

Added The Phantom Uragaan

Added Where Gods Fear to Tread

Cleaned up comments

Added quest restriction type enum

Changed custom quest drop table and fixed other supply crates

Fixed typo

### CODE REVIEW

**Comment by sepalani** [comment]
Could you please move all the quest related to code to a new file like `mh/quest_utils.py`? Then, you can import the quest_utils' constants in `constants`.

**Comment by kbarkevich** [comment]
> Could you please move all the quest related to code to a new file like `mh/quest_utils.py`? Then, you can import the quest_utils' constants in `constants`.

Done

File: mh/quest_utils.py [hunk 1]
```diff
@@ -0,0 +1,1564 @@
...
+
+QUEST_FLAG_UNK = 0b1000  # It really is unknown, it put a question mark icon on the monster/item box icon
+QUEST_FLAG_ARENA = 0b10000
+QUEST_FLAG_DOUBLE_MONSTER = 0b10000
+
+
+def pad(s, size, p=b'\0'):
+    data = bytearray(s + p * max(0, size-len(s)))
+    data[-1] = 0
+    return data
```
Comment (sepalani) [inline mh/quest_utils.py:14]: This function already exists in `mh/constants.py`, if you really need it you should move the one from the `mh.constants` module into a more relevant place like `other/utils.py` and import it rather than copy-paste it.

File: mh/quest_utils.py [hunk 2]
```diff
@@ -0,0 +1,1564 @@
+import struct
+
+
+LAYER_CHAT_COLORS = (0xbb3385ff, 0xffffffff, 0xffffffff)
```
Comment (sepalani) [inline mh/quest_utils.py:4]: This constant is unrelated to the quests and should be (re)moved.

File: mh/quest_utils.py [hunk 3]
```diff
@@ -0,0 +1,1564 @@
...
+    QUEST_LOCATION_TUNDRA = 4
+    QUEST_LOCATION_VOLCANO = 5
+    QUEST_LOCATION_GRT_DESERT = 6
+    QUEST_LOCATION_UW_RUIN = 7
+    QUEST_LOCATION_LAND_ARENA_1 = 8
+    QUEST_LOCATION_LAND_ARENA_2 = 9
+    QUEST_LOCATION_WATER_ARENA_1 = 10
+    QUEST_LOCATION_SACRED_LAND = 11
+    QUEST_LOCATION_WATER_ARENA_2 = 12
+location_type = Location_Type()
```
Comment (sepalani) [inline mh/quest_utils.py:32]: The comment can be removed and the parentheses as well. We usually don't use underscore (`_`) in class name. The class doesn't need to be instantiated, you can do it directly like that: `LocationType.QUEST_LOCATION_NONE`.

Example:
```python
class MyEnum:
    A = 42

print("The answer is {}".format(MyEnum.A))
```

File: mh/quest_utils.py [hunk 4]
```diff
@@ -0,0 +1,1564 @@
...
+
+class Monster():
```
Comment (sepalani) [inline mh/quest_utils.py:34]: Ditto.

File: mh/quest_utils.py [hunk 5]
```diff
@@ -0,0 +1,1564 @@
...
+    psc_wyv = 42
+    levthn = 43
+    elddrg = 44
+    lynian = 45
+    herbvr = 46
+    nptron = 47
+    fish = 48
+monster = Monster()
+
+class Quest_Rank_Type():
```
Comment (sepalani) [inline mh/quest_utils.py:86]: Ditto.

File: mh/quest_utils.py [hunk 6]
```diff
@@ -0,0 +1,1564 @@
...
+    none = 0
+    star_1 = 1
+    star_2 = 2
+    star_3 = 3
+    star_4 = 4
+    star_5 = 5
+    urgent = 6
+quest_rank_type = Quest_Rank_Type()
+
+class Items_Type():
```
Comment (sepalani) [inline mh/quest_utils.py:96]: Ditto.

File: mh/quest_utils.py [hunk 7]
```diff
@@ -0,0 +1,1564 @@
...
+    catalyst_jewel = 740
+    guts_jewel = 741
+    fortitude_jewel = 742
+    salvo_jewel = 743
+    fusillade_jewel = 744
+    dung_jewel = 745
+    torchlight_jewel = 746
+items_type = Items_Type()
+
+class Resources_Type():
```
Comment (sepalani) [inline mh/quest_utils.py:846]: Ditto.

File: mh/quest_utils.py [hunk 8]
```diff
@@ -0,0 +1,1564 @@
...
+    low_rank = 0x0000
+    high_rank = 0x0001
+    arena = 0x0002
+resources_type = Resources_Type()
+
+class Starting_Position_Type():
```
Comment (sepalani) [inline mh/quest_utils.py:852]: Ditto.

File: mh/quest_utils.py [hunk 9]
```diff
@@ -0,0 +1,1564 @@
...
+    camp = 0x00
+    random = 0x01
+    shrine = 0x02
+starting_position_type = Starting_Position_Type()
+
+class Quest_Restriction_Type():
```
Comment (sepalani) [inline mh/quest_utils.py:858]: Ditto.

**Review by sepalani — CHANGES_REQUESTED** [review]
I'm gonna let slide the other coding style and non-pythonic code issues as they are in a separated file. Otherwise, there a some points which need to be addressed and it should be good enough.

File: mh/quest_utils.py [hunk 10]
```diff
@@ -0,0 +1,1564 @@
+import struct
```
Comment (sepalani) [inline mh/quest_utils.py:1]: It's missing a licence header:
```python
#! /usr/bin/env python
# -*- coding: utf-8 -*-
"""Quest utils module.

    Monster Hunter 3 Server Project
    Copyright (C) 2022  NAME_TO_BE_CREDITED

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
```
Don't forget to change `NAME_TO_BE_CREDITED`

**Review by sepalani — CHANGES_REQUESTED** [review]
There is that too, now that I think about it.

File: mh/constants.py [hunk 11]
```diff
@@ -257,6 +257,9 @@ def slot(item, qty):
 OTHER_CAPCOM_ID = b"D9R7K4"
 OTHER_HUNTER_NAME = b"Drakea"
 
+SHIFT_OFFSET = 0
```
[the remark this answers is not in the document]
Comment (kbarkevich) [reply mh/constants.py:260]: This was a leftover from one of the other attempts at implementing event quests that got copied into my branch. Deleting it

File: mh/constants.py [hunk 12]
```diff
@@ -233,9 +236,6 @@ def slot(item, qty):
 
     return data
 
-
-LAYER_CHAT_COLORS = (0xbb3385ff, 0xffffffff, 0xffffffff)
```
[the remark this answers is not in the document]
Comment (kbarkevich) [reply mh/constants.py:237]: This was accidentally copied over from constants.py and has been returned there.

**Comment by kbarkevich** [comment]
Addressed all change requests

File: mh/constants.py [hunk 13]
```diff
@@ -257,6 +255,7 @@ def slot(item, qty):
 OTHER_CAPCOM_ID = b"D9R7K4"
 OTHER_HUNTER_NAME = b"Drakea"
 
+
```
Comment (sepalani) [inline mh/constants.py:258]: Is this newline really needed? I don't recall PEP8 to blame this line.

File: mh/constants.py [hunk 14]
```diff
@@ -283,43 +282,43 @@ def slot(item, qty):
     },
     0x06: {
         "version": 1,
-        "content": b"dummy_06\0"
-    },
+        "content": QUEST_EVENT_JUMP_FOUR_JAGGI
+    },
```
Comment (sepalani) [inline mh/constants.py:286]: The whitespace after the coma is still here and need to be removed.

File: mh/quest_utils.py [hunk 15]
```diff
@@ -0,0 +1,1567 @@
...
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import struct
+from other.utils import pad
+
+
+# LocationType enum
+class LocationType():
```
Comment (sepalani) [inline mh/quest_utils.py:27]: The parentheses are unnecessary and should be removed.

File: mh/quest_utils.py [hunk 16]
```diff
@@ -0,0 +1,1567 @@
...
+    QUEST_LOCATION_VOLCANO = 5
+    QUEST_LOCATION_GRT_DESERT = 6
+    QUEST_LOCATION_UW_RUIN = 7
+    QUEST_LOCATION_LAND_ARENA_1 = 8
+    QUEST_LOCATION_LAND_ARENA_2 = 9
+    QUEST_LOCATION_WATER_ARENA_1 = 10
+    QUEST_LOCATION_SACRED_LAND = 11
+    QUEST_LOCATION_WATER_ARENA_2 = 12
+
+class Monster():
```
Comment (sepalani) [inline mh/quest_utils.py:42]: Ditto.

File: mh/quest_utils.py [hunk 17]
```diff
@@ -0,0 +1,1567 @@
...
+    brt_wyv = 41
+    psc_wyv = 42
+    levthn = 43
+    elddrg = 44
+    lynian = 45
+    herbvr = 46
+    nptron = 47
+    fish = 48
+
+class QuestRankType():
```
Comment (sepalani) [inline mh/quest_utils.py:93]: Ditto.

File: mh/quest_utils.py [hunk 18]
```diff
@@ -0,0 +1,1567 @@
...
+    none = 0
+    star_1 = 1
+    star_2 = 2
+    star_3 = 3
+    star_4 = 4
+    star_5 = 5
+    urgent = 6
+
+class ItemsType():
```
Comment (sepalani) [inline mh/quest_utils.py:102]: Ditto.

File: mh/quest_utils.py [hunk 19]
```diff
@@ -0,0 +1,1567 @@
...
+    low_rank = 0x0000
+    high_rank = 0x0001
+    arena = 0x0002
+
+class StartingPositionType():
+    camp = 0x00
+    random = 0x01
+    shrine = 0x02
+
+class QuestRestrictionType():
```
Comment (sepalani) [inline mh/quest_utils.py:861]: Ditto for the parentheses.

**Review by sepalani — CHANGES_REQUESTED** [review]
Beside these nitpicks, LGTM.

File: mh/constants.py (hunk 13 above)
Comment (kbarkevich) [reply mh/constants.py]: Removed it

**Review by sepalani — APPROVED** [review]
LGTM regarding the code, untested.

**Comment by InusualZ** [comment]
Code LGTM for a first iteration. I would like to test this with the PAL version (I should be able to do it later today)

**Review by InusualZ — APPROVED** [review]
